feat: streamline TYPO3 non-composer mode configuration support

sbuerk opened the pull request:
<!-- 
  PR titles have very precise rules, please read 
  https://ddev.readthedocs.io/en/stable/developers/building-contributing/#open-pull-requests
-->

## The Issue

TYPO3 v12 changed the naming of the configuration files and also the place depending on the installation mode, which differs between composer and non-composer setups.

Since #4268 ddev already uses the correct files and places for composer based installations but not for non-composer mode setups, which includes also TYPO3 Core Development monorepo setups.

TYPO3 is going to remove the silent upgrade and also not writing the invalid `trustedHostsPattern` required for ddev. That means, that without the correct
pattern ddev setup for non-composer modes AND monorepo needs manual adjustmenst
and copy of the written file mainly because the default of typo3 no longer matches
the ddev variant. That means, the installer needs to be executed with `http` instead of `https`.

pkg/ddevapp/typo3.go` contains detection for TYPO3 v12 composer mode and TYPO3 legacy mode, using `app` in the method name which makes it hard to connect it with TYPO3 ecosystem known facts.

TYPO3 v12 changed the nameing of the main configuration file and the additional configuration file:

| TYPO3 <= v11                  | TYPO3 >= v12     |                    
|-------------------------------|------------------|
| `LocalConfiguration.php`      | `system.php`     |
| `AdditionalConfiguration.php` | `additional.php` |

Since TYPO3 v12 the place for these configuration files changed depending on the installation mode (composer or legacy mode):

| mode     | <= TYPO3 v11           | TYPO3 >= v12                  |
|----------|------------------------|-------------------------------|
| legacy   | `<docroot>/typo3conf/` | `<docroot>/typo3conf/system/` |
| composer | `<docroot>/typo3conf/` | `<project>/config/system/`    |

Note that in legacy mode `<project>` and `<docroot>` is the same folder,
where in composer mode `<docroot>` is usually `<project>/public/`

ddev respects this change only for TYPO3 v12 or higher in composer mode,
but not for the legacy mode and also not for the TYPO3 Core Development
mono repository.

That has been undetected so far because of two factors:

* many legacy setups may not be used with TYPO3 or developers has taken care of that manually anyway by disabling the ddev config writing for existing projects
* the fact that TYPO3 used to execute silent upgrade procedures also during installing new instances - which makes absolutly no sense and has been removed for TYPO3 v13 (main) and TYPO3 v12 [[1]](https://review.typo3.org/c/Packages/TYPO3.CMS/+/81793) along with writing the TYPO3 `trustedHostPattern` configuration to `.*.*`. That is also written by ddev to the `additional.php/AdditionalConfiguration.php`.

Because the old nameing is no longer migrated the web installer cannot be acces using `https://` and throwing exception in some cases related to the not suitable trustedHostPattern.

`trustedHostPattern` is used to verify hosts when reverse proxies are used, which is the case internally in ddev.

The issue is a combination of not updated detection (failed communication from TYPO3 to ddev) and the fact that TYPO3 removed unneeded code exection from the installation process and using more suitable `trustedHostPattern` by default which has hidden the issue away for quite some time.

[1] [https://review.typo3.org/c/Packages/TYPO3.CMS/+/81793](https://review.typo3.org/c/Packages/TYPO3.CMS/+/81793)

## How This PR Solves The Issue

This change renames all existing `isTypo3*()` function in `pkg/ddevapp/typo3.go` into more speaking function names and adds function comments to describe the background for the mode it detects. In the same manner a new mode detect function is added to detect TYPO3 v12 or higher in legacy mode properly.

The renamed `isTypo3App()` function has been used only as sub detection within `setTypo3SiteSettingsPaths()` and thus the renaming is okay so far. However,
the same function is used within the `pkg/ddevapp/apptypes.go` file as appTypeDetect function for the `nodeps.AppTypeTYPO3` type. Only checking for one "possible" type is not really usefull and thus the `isTypo3()` function is readded and now calls the different mode detection function in a chain to properly detect all possible TYPO3 setup flavours.

These changes takes care that the correct places and filenames are used for every possible case and ddev experiance is a breeze again like it should be. And not having to worry what to change or move or copy files arround while digging through support logs and chats.

To make that more clearer for developer and maintainers looking onto this code and not knowing these literally publicly known differences in the TYPO3 ecosystem and community,
the detection functions gets now some comments explaining the background a little bit
more in detail.

## Manual Testing Instructions

In all below descripted variants, instead of `typo3conf/LocalConfiguration.php` and `typo3conf/AdditionalConfiguration.php` the files `typo3conf/system/settings.php` and `typo3conf/system/additional.php` should be created and used.

That reads in ddev output like:

```terminal
TYPO3 does not seem to have been set up yet, missing settings.php (/home/sbuerk/work/TYPO3-Contribute/typo3conf/system/settings.php) 
Generating additional.php file for database connection. 
```

and using the installer works. Specially the monorepo with the change in TYPO3 no longer using the silentupgradeservice and not writing the ddev required trustedHostsPattern makes the difference.

### TYPO3 "legacy" (classic) mode, mono-repo (composer for dependencies, not the actual project!)

```terminal
git clone https://github.com/typo3/typo3.git ./
ddev config --project-name 't3c-main' --project-type 'typo3' --docroot '.' --database 'mariadb:10.11' --php-version '8.2' --webserver-type 'apache-fpm' --web-environment='TYPO3_CONTEXT=Development'
ddev start
# This is run locally, not within ddev by intent (otherwise would be docker-in-docker).
# Don't be disturbed by our main dispatcher called "runTests", it's not only for tests.
ddev composer install
ddev exec touch FIRST_INSTALL
ddev launch /typo3/install.php
```

### Legacy/Classic mode, NON-Composer

```terminal
# TYPO3 "legacy" (classic) mode, latest v13 release (NO composer)
wget --content-disposition https://get.typo3.org/13.3.0
tar xzf typo3_src-13.3.0.tar.gz
ln -s typo3_src-13.3.0 typo3_src && ln -s typo3_src/index.php index.php && ln -s typo3_src/typo3 typo3
ddev config --project-name 't3c-main' --project-type 'typo3' --docroot '.' --database 'mariadb:10.11' --php-version '8.2' --webserver-type 'apache-fpm' --web-environment='TYPO3_CONTEXT=Development'
ddev start
ddev exec touch FIRST_INSTALL
ddev launch /typo3/install.php

# TYPO3 "legacy" (classic) mode, stable v12 release (NO composer)
# differences to above only in version numbers and DDEV PHP-Version!
wget --content-disposition https://get.typo3.org/12.4.20
tar xzf typo3_src-12.4.20.tar.gz
ln -s typo3_src-12.4.20 typo3_src && ln -s typo3_src/index.php index.php && ln -s typo3_src/typo3 typo3
ddev config --project-name 't3c-main' --project-type 'typo3' --docroot '.' --database 'mariadb:10.11' --php-version '8.1' --webserver-type 'apache-fpm' --web-environment='TYPO3_CONTEXT=Development'
ddev start
ddev exec touch FIRST_INSTALL
ddev launch /typo3/install.php

# TYPO3 "legacy" (classic) mode, oldstable v11 release (NO composer)
# differences to above only in version number
wget --content-disposition https://get.typo3.org/11.5.39
tar xzf typo3_src-11.5.39.tar.gz
ln -s typo3_src-11.5.39 typo3_src && ln -s typo3_src/index.php index.php && ln -s typo3_src/typo3 typo3
ddev config --project-name 't3c-main' --project-type 'typo3' --docroot '.' --database 'mariadb:10.11' --php-version '8.1' --webserver-type 'apache-fpm' --web-environment='TYPO3_CONTEXT=Development'
ddev start
ddev exec touch FIRST_INSTALL
ddev launch /typo3/install.php
```

### TYPO3 composer mode (unchanged)

```terminal
# TYPO3 composer mode, latest v13 release
ddev config --project-name 't3c-main' --project-type 'typo3' --docroot 'public' --database 'mariadb:10.11' --php-version '8.2' --webserver-type 'apache-fpm' --web-environment='TYPO3_CONTEXT=Development'
ddev start
ddev composer create "typo3/cms-base-distribution:^13"
ddev exec touch public/FIRST_INSTALL
ddev launch /typo3/install.php

# TYPO3 composer mode, stable v12 release
# difference only in version numvers!
ddev config --project-name 't3c-main' --project-type 'typo3' --docroot 'public' --database 'mariadb:10.11' --php-version '8.1' --webserver-type 'apache-fpm' --web-environment='TYPO3_CONTEXT=Development'
ddev start
ddev composer create "typo3/cms-base-distribution:^12"
ddev exec touch public/FIRST_INSTALL
ddev launch /typo3/install.php

# TYPO3 composer mode, oldstable v11 release
# difference only in version numvers!
ddev config --project-name 't3c-main' --project-type 'typo3' --docroot 'public' --database 'mariadb:10.11' --php-version '8.1' --webserver-type 'apache-fpm' --web-environment='TYPO3_CONTEXT=Development'
ddev start
ddev composer create "typo3/cms-base-distribution:^11"
ddev exec touch public/FIRST_INSTALL
ddev launch /typo3/install.php
```

## Automated Testing Overview

<!-- Please describe the tests introduced by this PR, or explain why no tests are needed. -->

## Release/Deployment Notes

<!-- Does this affect anything else or have ramifications for other code? Does anything have to be done on deployment? -->

sbuerk commented:
Tested manually against monorepo main, and TYPO3 non-composer 13.2.1, 12.4.19 and 11.5.39 and with composer mode 13.2.1

sbuerk commented:
@rfay Is there something missing with my PR ? Can I provide additional info or things to get this into review process ?

rfay commented:
We'll get there. Just a little patience, thanks!

rfay commented:
Would you mind simplifying the manual testing? All it really needs is 

```
git checkout <something>
ddev config <perhaps-something>
```

or
```
wget <something> && untar
ddev config <perhaps-something>
```
right? 

If you can simplify those it will be easier to test. I'll test in a separate directory for each; don't have to re-use an existing directory.

rfay commented on pkg/ddevapp/typo3.go:
```diff
@@ -133,6 +133,10 @@ func setTypo3SiteSettingsPaths(app *DdevApp) {
 		settingsFileBasePath := filepath.Join(app.AppRoot, app.ComposerRoot)
 		settingsFilePath = filepath.Join(settingsFileBasePath, "config", "system", "settings.php")
 		localSettingsFilePath = filepath.Join(settingsFileBasePath, "config", "system", "additional.php")
+	} else if isTypo3Appv12OrHigher(app) {
```
I imagine this function needs some renaming to distinguish it from `isTypo3v12OrHigher` - what exactly is the difference with `isTypo3Appv12OrHigher` ? I assume "Typo3App" is somehow different from "Typo3", that should be explained somewhere and probably needs better naming.

sbuerk replied:
**isTypo3v12OrHigher**

TYPO3 composer mode, v12 or higher, which relevant because of changed configuration files and paths:

```
public/typo3conf/LocalConfiguration.php  -> config/system/settings.php
public/typo3conf/AdditionalConfiguration -> config/system/additional.php
```

Note that the files are moved out of the docroot (`public/*`) in composer mode.

Suggested rename: : `isTypo3v12OrHigherComposerMode`

Composer mode before TYPO3 v12 installed even in composer mode the public/typo3/sysext/folder ... function used the configured docroot (public/)
and thus handled TYPO3 composer mode before v12 with `isTypo3App`

**isTypo3App**

TYPO3 in non-composer mode (legacy mode) aka symlink based setup. The TYPO3 monorepo (core development) has the "typo3/sysext/*" folder like a legacy setup.

Here a simple check has been used, simply looking that `typo3/` folder exists
in the doc root (TYPO3 core files). 

This check is still correct for non-composer mode and also monorepo (core development) - and kind of worked.

Before TYPO3 v12 the configuration files resides in:

```
public/typo3conf/LocalConfiguration.php
public/typo3conf/AdditionalConfiguration
```

which has changed starting with TYPO3 v12. See `isTypo3Appv12OrHigher` below.

Suggested rename: `isTypo3v11OrLowerComposerAndNonComposerMode`

**isTypo3Appv12OrHigher**

I cloned the **isTypo3App*  function and added the Typo3Version class check to it, which is only available since TYPO3 v12 to use the new placed for configuration files in TYPO3 v12 or newer, not covered by boath methods above:

```
public/typo3conf/LocalConfiguration.php  -> public/typo3conf/system/settings.php
public/typo3conf/AdditionalConfiguration -> public/typo3conf/system/additional.php
```

Suggested rename: isTypo3v12OrHigherLegacyMode


Would the renaming be okay for you ?

rfay replied:
`isTypo3v12OrHigherLegacyMode()` do you have docs for "legacy mode"? It seems like a reasonable approach in that case, but it needs a link to "legacy mode" docs. 

I have to admit that I thought T3 was always done with composer!

fe-hicking replied:
The "legacy mode" (actually: "classic" mode) is described here:

https://docs.typo3.org/m/typo3/tutorial-getting-started/main/en-us/Installation/LegacyInstallation.html#legacyinstallation

It is used for core dev installs (also referred to as "monorepo" install):

https://docs.typo3.org/m/typo3/guide-contributionworkflow/main/en-us/Quickstart/Index.html#quickstart

Sorry for that being confusing. For a 26year old system things may get cluttered.

What Stefan is proposing here is backed by the core team, in this matter currently represented by Stefan Bürk, Benjamin Franzke and me too. 😅 (ref: https://typo3.org/community/teams/typo3-development, section "framework mergers").

sbuerk commented:
> Would you mind simplifying the manual testing? All it really needs is
> 
> ```
> git checkout <something>
> ddev config <perhaps-something>
> ```
> 
> or
> 
> ```
> wget <something> && untar
> ddev config <perhaps-something>
> ```
> 
> right?
> 
> If you can simplify those it will be easier to test. I'll test in a separate directory for each; don't have to re-use an existing directory.

Don't get what you mean. Is your question related towards the mkdir/remove ? I used the clone of the monorepo because of the pending (now merged) change - which revealed that the configuration files for non-composer mode were wrong (since v12) but "silently covered" by an upgrade we no longer execute on installation (new setups should not need an configuration migration).

Will recheck the steps tomorrow in a break and reduce it / update them due to changed core patch (v13 / v12 ).

> We'll get there. Just a little patience, thanks!

Sorry - I did not wanted to pressure you or the team. I only wanted to ask if I have missed to provide something, and due to being in the "hot phase" for TYPO3 v13 now (feature freeze release coming up) I miss phrased the question. (Just want to provide as much as needed from our/myside to minimuze the work on your side).

Thanks for the good work :heart: .

rfay commented:
> Don't get what you mean. Is your question related towards the mkdir/remove ? 

Yes, I'm asking for a simple start. It should only take a couple of lines to demonstrate each thing you have, not a monster that starts with `rm -r`. And the `&&` are not required, just give separate lines for each action. 

It's important to express what you're actually trying to accomplish, not just show a lot of lines of actions. For each of them, please say what case they're trying to accomplish, and check them out or download them into a different directory, like the various DDEV quickstarts do, https://ddev.readthedocs.io/en/stable/users/quickstart/

Remember that we DDEV maintainers don't have the intimate understanding of TYPO3 that you do, so we need to understand what's being attempted in each stanza and why.

Also, please make clear what TYPO3 maintainers will review and approve this, thanks!

sbuerk commented:
> > Don't get what you mean. Is your question related towards the mkdir/remove ?
> 
> Yes, I'm asking for a simple start. It should only take a couple of lines to demonstrate each thing you have, not a monster that starts with `rm -r`. And the `&&` are not required, just give separate lines for each action.
> 
> It's important to express what you're actually trying to accomplish, not just show a lot of lines of actions. For each of them, please say what case they're trying to accomplish, and check them out or download them into a different directory, like the various DDEV quickstarts do, https://ddev.readthedocs.io/en/stable/users/quickstart/
> 
> Remember that we DDEV maintainers don't have the intimate understanding of TYPO3 that you do, so we need to understand what's being attempted in each stanza and why.
> 
> Also, please make clear what TYPO3 maintainers will review and approve this, thanks!

Thanks. I guess I will update with @garvinhicking suggestion shortly.

> Also, please make clear what TYPO3 maintainers will review and approve this, thanks!

I'm core maintainer myself, so is Garvin - but I will ping some to veryify this after the test instruction change.

rfay commented:
> I'm core maintainer myself, so is Garvin - but I will ping some to veryify this after the test instruction change.

That's fantastic, and of course thanks for working on this. 

Can you say a little about why the directory structure is different when non-composer mode? That's a bit of a surprise.

garvinhicking commented:
The main reason I believe is that composer projects have a distinction for their "document root" as a subdirectory. And that subdirectory (web-dir) can be specified.

In the old way we have, because composer wasn't invented then, the root directory is the documentroot, so every subdirectory of the project lives in the document root and a special directory like "typo3temp". These are gated by htaccess controls only.

Maybe Stefan has some more reasoning, but mainly it's: "because it's always been like this".

TYPO3 has about roughly 50% composer adoption. The other 50% of websites powered by TYPO3 do NOT use composer.

An extension developer will use a composer installation of TYPO3; but for all us core developers and contributors, we need to use the monorepo-non-composer-way. Because the monorepo is split into the composer packages and only by the current structure it allows to power both composer and non-composer way.

rfay commented:
> The main reason I believe is that composer projects have a distinction for their "document root" as a subdirectory. And that subdirectory (web-dir) can be specified.

DDEV is perfectly fine with the docroot being the project root. Is it possible that the existing code is just ignoring that fact?

garvinhicking commented:
TYPO3 internally uses the directory "typo3conf" for the classic/legacy/non-composer way, but "config" for the composer one. So this is not something that DDEV can influence,  but it would need to respect that. That's what the patch is actually about, to see whether it needs to put files into "config" or into "typo3conf".

TYPO3 previously silently moved the files to the right locations if it found a mismatch. This is no longer performed in TYPO3 with the next release, so now applications setting up TYPO3 (in this case DDEV) need to adapt to write the proper location now.

The patch to change this has been merged with https://review.typo3.org/c/Packages/TYPO3.CMS/+/81793 and will be part of the next TYPO3 v12 and TYPO3 v13 release. Anyone using DDEV together with that new TYPO3 version either needs to:

* upgrade to a DDEV version that includes Stefans patch to DDEV or
* manually place the DDEV created files into the right directory

I don't want to cross-fire Stefans intent, but I hope this clears things up a bit more.

rfay commented:
When do we expect the new TYPO3 releases?

garvinhicking commented:
Next v12: October 8th
Next v13: September 17th (next week)

rfay commented:
5 days from now seems exotic for a new DDEV release. Oct 8 less exotic. 

We sure do appreciate you maintaining this and paying attention to everything! Of course fantastic Gilbertsoft used to do that for everything and he disappeared :(

garvinhicking commented:
Oh, don't worry, we're not on a hurry. It only affects a very limited set of users. But we want to take care of it. You do a great job, and we'll do it according to your schedule.  

(Sadly I don't know about the whereabouts of gilbertsof but I also appreciate everything he contributed)

sbuerk commented:
> Can you say a little about why the directory structure is different when non-composer mode? That's a bit of a surprise.

Legacy mode hat the index.php in the root and the typo3, typo3conf folder layout since ages.
With the rise of composer and the integration into the core, and having a "composer" mode the layout changed to have the vendor folder and configuration stuff outside of the docroot, thus having a public/ folder. which is already the case since quite some time.

With v12 and typo3/cms-composer-installers extensions are no longer installed in public/typo3conf/ext but kept in vendor/ - which symlinking resources folders to public/_assets.

That all works pretty fine with ddev.

However, with TYPO3 v12 writing configuration files has changed - which has been addressed only for composer mode.

For non-composer mode which still makes nearly 50% of TYPO3 installations this has not been addressed. ddev writes fro legacy mode to the places and filenames which were valid before TYPO3 v12.

TYPO3 had code in the installation process and setup command executing silent upgrade things even on "new" installations which "converted" the "AdditionalConfiguration.php" written by ddev to the new variant - but ddev continued working with the file no longer regonized by TYPO3.

Recently, that code has been removed (even backported to v12) which means, people trying to use ddev and TYPO3 in legacy mode the will get issues due to having not a correct trustHostPattern (TYPO3 configuration) written (TYPO3 no longer does this, and the ddev file will not be migrated anymore) - and if the installer is used it must be done either with http:// or inbetween after errors the trustPattern changed.

TYPO3 Core Development monorepo is literally a "legacy" mode installation in the root, beside having a "root composer.json" - and ddev user trying to contribute have now the same errors like the users of legacy mode.

That popped up recently on our side while already beeing wring since the TYPO3 v12 release (for non-composer mode) installations.

As we have feature freeze tomorrow (not LTS) release and I was quite busy for that, I will update the PR and code the next days.

I hope that explanation gives some background/insight.

> 5 days from now seems exotic for a new DDEV release. Oct 8 less exotic.

Correct - but don't let you be forced to make a release. We try to provide this now so you have not to deal with the why and then. And the release is coming when you feel condifent to release a new version. Point. Nothing to add here.

I myself do not use ddev at all and thus never dedected that, it just came up while testing the core change using ddev (installed it for that).


> We sure do appreciate you maintaining this and paying attention to everything! Of course fantastic Gilbertsoft used to do that for everything and he disappeared :(

I guess we will have now a closer eye if we are doing things which influences ddev or ddev users wanting to use it with TYPO3.

I know that gilbertsoft did that in the past - but disappeared.

I guess that most people using ddev and TYPO3 uses composer mode - but neverless I say we should make the changes so it works - and if not for project, at least for developers wanting to contribute to the TYPO3 core.

Can only repeat what garvin said - you are doing a great and amazing job and all my collegues are happy about ddev and it is a game change. Keep it up. And we try to help out if we see something ;)

rfay commented:
Thanks for the work on this. It doesn't build at this time. Please build it locally and verify behavior. Thanks!

sbuerk commented:
> Thanks for the work on this. It doesn't build at this time. Please build it locally and verify behavior. Thanks!

@rfay Sorry, I did not managed to get it locally buildable at all. Looked through the README.md, CONTRIBUTING.md and also  the building-contributing.md but I'm kind of lost on that. Checked the pipeline and hope fixed it manually now.

Maybe you have some time the next days and we can look into making a setup working and how to do it - if further pushes are needed. I hate it when I cannot build and check things first locally specially if the required pipeline are so long taking and expensive. Would appreciate that and maybe I will to the one or other adjustment in the future (for TYPO3).

I'm not a daily go developer, did that for smaller things and private toolings which is quite simple compared to the complex codebase of ddev (for me) and therefore not that experienced with such golang opensource projects. Just trying to do my best to contribute and help out.

I searched if some entry in a changelog needs to be provided with the change anyway - the version-history.md looks kind of already released onces. Did I missed here something ?

I have updated the PR topic comment with the instruction and also adjusted the description of it. Not sure how golang developer stands to comments, but I hope the code comments may help also in future when someone deals with these TYPO3 detection functions.

sbuerk commented:
Would also welcome some help or pointings how to add tests (apptype) for these different setup modes.

rfay commented:
Thanks! I know how difficult it is to navigate a different world. 

To build DDEV you normally just need to check it out and use `make`. go has to be installed, but that's usually it. 

Or just use [Open in Gitpod](https://ddev.readthedocs.io/en/stable/developers/building-contributing/#open-in-gitpod) for a pre-set-up environment. 

[Setting up a go development environment](https://ddev.com/blog/setting-up-a-go-development-environment/) may help too, with a training video. 

I'm happy to help any time to see if we can get over the hump, but if you can't manually test, then none of us will have any confidence in the result. Ping me here, ping me on Discord or TYPO3 slack. Make an appointment for a screenshare session at https://cal.com/randyfay/30min

Most of all, thanks for caring for this!

rfay commented:
And due to your fix, your build, which is now successful, is available for download and manual testing at https://github.com/ddev/ddev/pull/6525#issuecomment-2332013047

sbuerk commented:
> Thanks! I know how difficult it is to navigate a different world.
> 
> To build DDEV you normally just need to check it out and use `make`. go has to be installed, but that's usually it.
> 
> Or just use [Open in Gitpod](https://ddev.readthedocs.io/en/stable/developers/building-contributing/#open-in-gitpod) for a pre-set-up environment.
> 
> [Setting up a go development environment](https://ddev.com/blog/setting-up-a-go-development-environment/) may help too, with a training video. There are a number of others at https://ddev.com/blog/category/training/ . And some more recordings that don't show up there, https://ddev.com/blog/contributor-training/
> 
> I'm happy to help any time to see if we can get over the hump, but if you can't manually test, then none of us will have any confidence in the result. Ping me here, ping me on Discord or TYPO3 slack. Make an appointment for a screenshare session at https://cal.com/randyfay/30min
> 
> Most of all, thanks for caring for this!

Will do the manual verification again tomorrow and will also try to get @garvinhicking into it to do the same. Will ping you on that.

Adding additional automatic tests I guess I need more time to prepare that. When I do that I would create something which can update things etc in some repository (opensource) so the TYPO3 community can also help on updateing at least the testpackge generation thingy. Will see if I can manage that the next days. At list I try to great one manual forst for v13 or v12 as a starting point - just read through the code and I think I have a good guess now how this should work and try to start with a manual package first before (semi) automating it. 

Will ping you at some point to chat about the testpackages around.

rfay commented:
Tomorrow's Contributor training is about DDEV's automated tests. You're sure invited! https://www.meetup.com/ddev-events/events/302404699/?eventOrigin=group_upcoming_events

sbuerk commented:
# Result Overview

| mode     | TYPO3 | release | pr |
|----------|-------|---------|----|
| legacy   | v13   | ❌       | ✅  |
| mono     | v13   | ❌       | ✅  |
| composer | v13   | ✅       | ✅  |
| legacy   | v12   | ❌       | ✅  |
| mono     | v12   | ❌       | ✅  |
| composer | v12   | ✅       | ✅  |
| mono     | v11   | ✅       | ✅  |

> Note that mono repository is basically a legacy installation

# Test execution

<details>
  <summary>TYPO3 v13 - legacy mode</summary>

  ```bash
  # fetch release package, deflate it and prepare legacy mode symlink structure
  wget --content-disposition https://get.typo3.org/13.3.0
  tar xzf typo3_src-13.3.0.tar.gz
  ln -s typo3_src-13.3.0 typo3_src
  ln -s typo3_src/index.php index.php
  ln -s typo3_src/typo3 typo3
  
  # configure ddev
  ddev config \
      --project-name 't3c-v13-legacy' \
      --project-type 'typo3' \
      --docroot './' \
      --database 'mariadb:10.11' \
      --php-version '8.2' \
      --webserver-type 'apache-fpm' \
      --web-environment='TYPO3_CONTEXT=Development'
      
  # startup preparation    
  ddev start
  
  # provide TYPO3 installer flag file
  ddev exec touch FIRST_INSTALL
  
  # verify installer is accessible
  if [[ "$( curl -s "https://t3c-v13-legacy.ddev.site/typo3/install.php" )" == *"Installing TYPO3 CMS"* ]]; then echo "check passed"; else echo "check failed"; fi
  
  # verify extra configuration is in the expected path and has correct naming
  if [ -f "typo3conf/system/additional.php" ]; then echo "'typo3conf/system/additional.php' found"; else echo "'typoconf/system/additional.php' not found"; fi
  
  # expected output
  # Verify installer is accessible check passed
  # 'typo3conf/system/additional.php' found    
  ```
</details>

<details>
  <summary>TYPO3 v13 - mono repository</summary>

  ```bash
  # check out the TYPO3 core mono repository
  git clone https://github.com/TYPO3/typo3.git
  
  # configure ddev 
  ddev config \
      --project-name 't3c-v13-mono' \
      --project-type 'typo3' \
      --docroot './' \
      --database 'mariadb:10.11' \
      --php-version '8.2' \
      --webserver-type 'apache-fpm' \
      --web-environment='TYPO3_CONTEXT=Development'
      
  # startup preparation
  ddev start && ddev composer install
  
  # provide TYPO3 installer flag file
  ddev exec touch FIRST_INSTALL \
  
  # verify installer is accessible
  if [[ "$( curl -s "https://t3c-v13-mono.ddev.site/typo3/install.php" )" == *"Installing TYPO3 CMS"* ]]; then echo "check passed"; else echo "check failed"; fi
  
  # verify extra configuration is in the expected path and has correct naming
  if [ -f "typo3conf/system/additional.php" ]; then echo "'typo3conf/system/additional.php' found"; else echo "'typoconf/system/additional.php' not found"; fi
  
  # expected output
  # Verify installer is accessible check passed
  # 'typo3conf/system/additional.php' found   
  ```
</details>

<details>
  <summary>TYPO3 v13 - composer mode</summary>

  ```bash
  # configure ddev
  ddev config \
      --project-name 't3c-v13-composer' \
      --project-type 'typo3' \
      --docroot 'public/' \
      --database 'mariadb:10.11' \
      --php-version '8.2' \
      --webserver-type 'apache-fpm' \
      --web-environment='TYPO3_CONTEXT=Development'
      
  # install TYPO3 using base composer distribution package    
  ddev composer create "typo3/cms-base-distribution:^13"
  
  # provide TYPO3 installer flag file
  ddev exec touch public/FIRST_INSTALL \

  # verify installer is accessible
  if [[ "$( curl -s "https://t3c-v13-composer.ddev.site/typo3/install.php" )" == *"Installing TYPO3 CMS"* ]]; then echo "check passed"; else echo "check failed"; fi
  
  # verify extra configuration is in the expected path and has correct naming
  if [ -f "config/system/additional.php" ]; then echo "'config/system/additional.php' found"; else echo "'config/system/additional.php' not found"; fi
  
  # expected output
  # Verify installer is accessible check passed
  # 'config/system/additional.php' found
  ```
</details>

<details>
  <summary>TYPO3 v12 - legacy mode</summary>

  ```bash
  # fetch release package, deflate it and prepare legacy mode symlink structure
  wget --content-disposition https://get.typo3.org/12.4.20
  tar xzf typo3_src-12.4.20.tar.gz
  ln -s typo3_src-12.4.20 typo3_src
  ln -s typo3_src/index.php index.php
  ln -s typo3_src/typo3 typo3
  
  # configure ddev
  ddev config \
      --project-name 't3c-v12-legacy' \
      --project-type 'typo3' \
      --docroot './' \
      --database 'mariadb:10.11' \
      --php-version '8.1' \
      --webserver-type 'apache-fpm' \
      --web-environment='TYPO3_CONTEXT=Development'
            
  ddev start
  
  # provide TYPO3 installer flag file
  ddev exec touch FIRST_INSTALL
  
  # verify installer is accessible
  if [[ "$( curl -s "https://t3c-v12-legacy.ddev.site/typo3/install.php" )" == *"Installing TYPO3 CMS"* ]]; then echo "check passed"; else echo "check failed"; fi
  
  # verify extra configuration is in the expected path and has correct naming
  if [ -f "typo3conf/system/additional.php" ]; then echo "'typo3conf/system/additional.php' found"; else echo "'typoconf/system/additional.php' not found"; fi
  
  # expected output
  # Verify installer is accessible check passed
  # 'typo3conf/system/additional.php' found   
  ```
</details>

<details>
  <summary>TYPO3 v12 - mono repository</summary>

  ```bash
  # check out the TYPO3 core mono repository on v12 development branch
  git clone https://github.com/TYPO3/typo3.git .
  git checkout 12.4
  
  # configure ddev
  ddev config \
      --project-name 't3c-v12-mono' \
      --project-type 'typo3' \
      --docroot './' \
      --database 'mariadb:10.11' \
      --php-version '8.1' \
      --webserver-type 'apache-fpm' \
      --web-environment='TYPO3_CONTEXT=Development'
      
  # startup preparation    
  ddev start && ddev composer install
  
  ddev exec touch FIRST_INSTALL \
  
  # verify installer is accessible
  if [[ "$( curl -s "https://t3c-v12-mono.ddev.site/typo3/install.php" )" == *"Installing TYPO3 CMS"* ]]; then echo "check passed"; else echo "check failed"; fi
  
  # verify extra configuration is in the expected path and has correct naming
  if [ -f "typo3conf/system/additional.php" ]; then echo "'typo3conf/system/additional.php' found"; else echo "'typoconf/system/additional.php' not found"; fi 
  ```
</details>

<details>
  <summary>TYPO3 v12 - composer mode</summary>

  ```bash
  # configure ddev
  ddev config \
      --project-name 't3c-v12-composer' \
      --project-type 'typo3' \
      --docroot 'public/' \
      --database 'mariadb:10.11' \
      --php-version '8.1' \
      --webserver-type 'apache-fpm' \
      --web-environment='TYPO3_CONTEXT=Development'
      
  # install TYPO3 using base composer distribution package      
  ddev composer create "typo3/cms-base-distribution:^12"
  
  # provide TYPO3 installer flag file
  ddev exec touch public/FIRST_INSTALL \

  # verify installer is accessible
  if [[ "$( curl -s "https://t3c-v12-composer.ddev.site/typo3/install.php" )" == *"Installing TYPO3 CMS"* ]]; then echo "check passed"; else echo "check failed"; fi
  
  # verify extra configuration is in the expected path and has correct naming
  if [ -f "config/system/additional.php" ]; then echo "'config/system/additional.php' found"; else echo "'config/system/additional.php' not found"; fi
  
  # expected output
  # Verify installer is accessible check passed
  # 'config/system/additional.php' found  
  ```
</details>

<details>
  <summary>TYPO3 v12 - mono repository</summary>

  ```bash
  # check out the TYPO3 core mono repository on v12 development branch
  git clone https://github.com/TYPO3/typo3.git
  git checkout 11.5
  
  # configure ddev
  ddev config \
      --project-name 't3c-v11-mono' \
      --project-type 'typo3' \
      --docroot './' \
      --database 'mariadb:10.11' \
      --php-version '7.4' \
      --webserver-type 'apache-fpm' \
      --web-environment='TYPO3_CONTEXT=Development'
      
  # startup preparation   
  ddev start && ddev composer install
  
  # provide TYPO3 installer flag file
  ddev exec touch FIRST_INSTALL
  
  # verify installer is accessible
  if [[ "$( curl -s "https://t3c-v11-mono.ddev.site/typo3/install.php" )" == *"Installing TYPO3 CMS"* ]]; then echo "check passed"; else echo "check failed"; fi
  
  # verify extra configuration is in the expected path and has correct naming
  if [ -f "typo3conf/AdditionalConfiguration.php" ]; then echo "'typo3conf/AdditionalConfiguration.php' found"; else echo "'typoconf/AdditionalConfiguration.php' not found";
  
  # expected output
  Verify installer is accessible check passed
  'typo3conf/AdditionalConfiguration.php' found
  ```    

</details>

stasadev reviewed (approved):
I tested all the instructions provided and everything works as expected.

The `git clone` for the mono repository v11 and v13 was missing a dot, which at first confused me, but then I noticed it in v12 and everything worked, I updated the instructions.

Thank you @sbuerk !

rfay reviewed (approved):
This is great. Thanks for all the care and tender loving care! 

A couple of thoughts:

* The Drupal core devs have the same exact problem between composer and git, and there are a couple of competing approaches they use which might be useful. https://github.com/joachim-n/drupal-core-development-project and a DDEV add-on, https://github.com/justafish/ddev-drupal-core-dev (which seems to be unmaintained at this point). You may be interested in a DDEV add-on for this situation too.
* You don't need `--project-name` in most of these situations, the project will take the name of the directory it's in. 
* That's a lot of permutations of install techniques!!!

Again, thanks for taking care of TYPO3 and caring for DDEV's implementation so that TYPO3 folks can have an easy time.

sbuerk commented:
> I tested all the instructions provided and everything works as expected.
> 
> The `git clone` for the mono repository v11 and v13 was missing a dot, which at first confused me, but then I noticed it in v12 and everything worked, I updated the instructions.
> 
> Thank you @sbuerk !

Thanks - that happend while converting my `single command chains` to dedicated command per line. Sorry for that. Will improve that for the future.

sbuerk commented:
> This is great. Thanks for all the care and tender loving care!
> 
> A couple of thoughts:
> 
> * The Drupal core devs have the same exact problem between composer and git, and there are a couple of competing approaches they use which might be useful. https://github.com/joachim-n/drupal-core-development-project and a DDEV add-on, https://github.com/justafish/ddev-drupal-core-dev (which seems to be unmaintained at this point). You may be interested in a DDEV add-on for this situation too.
> * You don't need `--project-name` in most of these situations, the project will take the name of the directory it's in.
> * That's a lot of permutations of install techniques!!!
> 
> Again, thanks for taking care of TYPO3 and caring for DDEV's implementation so that TYPO3 folks can have an easy time.

Maybe we can have a call about that. I do not get how a addon would help with the fact that `ddev` have written incorrect additional configuration files in incorrect places.

Monorepo (dev) is basically a legacy installation. For core development core related scripts should be executed the contained `Build/Scripts/runTests.sh` script anyway outside ddev. We only need to take care that the configuraiton files are written correctly.

I added `--project-name` option to ensure correct names so the included verify scripts using curl works and did not needed adjustments (copy & paste should simply work). I know that it takes the folder, but omiting a mkdir statemant the verify steps would eventually call the wrong project url.

Will create one or two additional PR the next week (minor ones) addressing the one or other thing. Beside that I already started working on `test package` creating repository and detected that we need to fix our TYPO3 setup command to allow omitting the database parameters when ddev provided the additional configuration in the right place.

Thanks guy for this great tool and maintaining it for the broader (PHP) development community!

sbuerk commented on pkg/ddevapp/typo3.go:
```diff
@@ -125,32 +125,68 @@ func getTypo3Hooks() []byte {
 	var settingsFilePath, localSettingsFilePath string
 
-	if isTypo3v12OrHigher(app) {
+	if isTypo3ComposerV12OrHigher(app) {
```
Will eventually create a follow up for that. Thanks for mentioning it !

rfay commented:
I've spent a little time looking at 
* https://github.com/ddev/ddev-drupal-contrib
* https://github.com/joachim-n/drupal-core-development-project
* https://github.com/joachim-n/drupal-project-contrib-development

They work on the same general class of problem that you're working on here. Drupal core has always been a git clone with a required `composer install`, but ended up being different in configuration than any of the "real" user installs, which have the docroot in `web`. 

The joachim-n repositories set up a scaffolding where the core is checked out into a subdirectory, then other things can be added on to it, and it ends up behavior more like a real user. 

Just commenting on those in passing, as they might be interesting to you.

If you want to chat about those or other things, feel free to create a meeting any time at https://cal.com/randyfay/30min

Thanks!

garvinhicking commented:
@rfay We do have something along these lines: https://github.com/ochorocho/tdk (personally, I actually use this - then I can work with normal composer installation mode)

But that's a bit "work in progress" and since other matters are more pressing, it's not really a focus at this time. I agree though there are a lot of opportunities to improve this, alas so much to do and so little time :-D
